File: 0086-partition-list/0086-partition-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        dummy1, dummy2 = ListNode(0), ListNode(0)
        first, second = dummy1, dummy2
        while head:
            if head.val < x:
                first.next = head
                first = head
            else:
                second.next = head
                second = head
            head = head.next
        first.next = dummy2.next
        second.next = None
        return dummy1.next

File: 0086-partition-list/0086-partition-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} x
 * @return {ListNode}
 */
var partition = function(head, x) {
  const dummy1 = new ListNode(0);
  const dummy2 = new ListNode(0);
  let [p1, p2] = [dummy1, dummy2];
  
  while (head) {
    if (head.val < x) {
      p1.next = head;
      p1 = head;
    } else {
      p2.next = head;
      p2 = head;
    }
    
    head = head.next;
  }
  
  p1.next = dummy2.next;
  p2.next = null;
  
  return dummy1.next;
};